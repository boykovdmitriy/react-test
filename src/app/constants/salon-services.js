import keyMirror from 'fbjs/lib/keyMirror';

export const SalonServices = keyMirror({
    SALON_SERVICES_REQUEST: undefined,
    SALON_SERVICES_SUCCESS: undefined,
    SALON_SERVICES_FAILURE: undefined,
    SALON_SERVICES_CANCEL: undefined,
});