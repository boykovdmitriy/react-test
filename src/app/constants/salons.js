import keyMirror from 'fbjs/lib/keyMirror';

export const Salons = keyMirror({
    SALON_REQUEST: undefined,
    SALON_SUCCESS: undefined,
    SALON_FAILURE: undefined,
    SALON_CANCEL: undefined,
    SALONS_REQUEST: undefined,
    SALONS_SUCCESS: undefined,
    SALONS_FAILURE: undefined,
    SALONS_CANCEL: undefined,
});