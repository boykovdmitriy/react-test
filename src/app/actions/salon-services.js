import {SalonServices} from "constants";

export function salonServicesRequest(salonId) {
    return {
        type: SalonServices.SALON_SERVICES_REQUEST,
        payload: {salonId},
    };
}