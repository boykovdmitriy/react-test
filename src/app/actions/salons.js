import {Salons} from "constants";

export function salonsRequest(page, pageSize) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {page, pageSize},
    };
}

export function salonRequest(salonId) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {salonId},
    };
}