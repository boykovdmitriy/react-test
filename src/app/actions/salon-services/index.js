import {SalonServices} from "constants/salon-services";

export function salonServicesRequest(salonId) {
    return {
        type: SalonServices.SALON_SERVICES_REQUEST,
        payload: {salonId},
    };
}