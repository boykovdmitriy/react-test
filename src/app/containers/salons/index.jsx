import React from "react";
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import {InfiniteScroll} from 'components/infinite-scroll';

import {ProfileImage} from "components/profile-image";
import {salonsRequest} from "actions";
import {Table, TableColumn, TableRow} from "components/table";
import {Spinner} from 'components/spinner';

class Salons extends React.Component {
    constructor(props) {
        super(props);
        this.pageSize = 25;
    }

    componentDidMount() {
        const {loadedUntilPage} = this.props;
        if (!loadedUntilPage) {
            this.loadNextPage();
        }
    }

    loadNextPage() {
        const {fetchSalons, loadedUntilPage, total} = this.props;
        const nextPage = loadedUntilPage ? loadedUntilPage + 1 : 1;
        if (!total || nextPage <= total) {
            fetchSalons(nextPage, this.pageSize);
        }
    }

    handleScroll() {
        this.loadNextPage();
    }

    render() {
        const {isLoading, salons} = this.props;
        return (
            <section>
                <InfiniteScroll handleScroll={() => this.handleScroll()}>
                    <Table headers={['Name', 'Website', 'Image']}>
                        {salons.length !== 0 && salons.map(x => (
                            <TableRow key={x.id}>
                                <TableColumn>{x.name}</TableColumn>
                                <TableColumn>{!!x.website && <a href={x.website}>link</a>}</TableColumn>
                                <TableColumn>{
                                    !!x.profile_image_urls &&
                                    <ProfileImage thumb={x.profile_image_urls.thumb}/>
                                }</TableColumn>
                                <TableColumn><Link to={`salon/${x.id}`}>Details</Link></TableColumn>
                            </TableRow>
                        ))}
                    </Table>
                </InfiniteScroll>
                {isLoading && <Spinner/>}
            </section>
        );
    }
}

const mapStateToProps = state => ({
    isLoading: state.salons.list.isLoading,
    salons: state.salons.list.items,
    totalPage: state.salons.list.total,
    loadedUntilPage: state.salons.list.page
});
const mapDispatchToProps = dispatch => ({
    fetchSalons: (page, pageSize) => dispatch(salonsRequest(page, pageSize))
});
export default connect(mapStateToProps, mapDispatchToProps)(Salons);



