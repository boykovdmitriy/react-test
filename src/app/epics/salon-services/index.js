import {ofType} from 'redux-observable';
import {switchMap} from 'rxjs/operators'
import {SalonServices} from 'constants/salon-services';
import {rxAjax} from 'utils';

export function salonServicesRequest(action$) {
    return action$.pipe(ofType(SalonServices.SALON_SERVICES_REQUEST),
        switchMap((action) => {
            const endpoint = `http://staging.salony.com/v5/salons/${action.payload.salonId}/services`;

            /* istanbul ignore next */
            return rxAjax({
                endpoint,
                method: 'GET',
            })
                .map(data => ({
                    type: SalonServices.SALON_SERVICES_SUCCESS,
                    payload: data.response,
                }))
                .catch(error => ({
                    type: SalonServices.SALON_SERVICES_FAILURE,
                    payload: {message: error.error, status: error.status},
                }));
        })
    );
}
