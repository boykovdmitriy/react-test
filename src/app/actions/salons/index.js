import {Salons} from "constants/salons";

export function salonsRequest(page, pageSize) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {page, pageSize},
    };
}

export function salonRequest(salonId) {
    return {
        type: Salons.SALON_REQUEST,
        payload: {salonId},
    };
}