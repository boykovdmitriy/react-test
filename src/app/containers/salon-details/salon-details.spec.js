import React from 'react';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import ConnectedSalonDetails, {SalonDetails} from './index'
import {SalonServices} from "constants/salon-services";
import {Salons} from "constants/salons";

const initialState = {
    salons: {
        item: {
            isLoading: false,
            salon: {
                id: 12,
                name: 'TestName',
                description: 'TestDescription',
                hours: {
                    fri: {to: "20:00", from: "07:00"},
                    mon: {to: "20:00", from: "07:00"},
                },
                images: [
                    {
                        id: 1,
                        image_urls: {
                            medium: 'http://s3-eu-west-1.amazonaws.com/staging-salony/image/salons/files/000/003/623/medium/nail-art-2392-250x250.jpg?1507118077'
                        }
                    }
                ]
            }
        }
    },
    salonServices: {
        isLoading: false,
        services: [{
            id: 1,
            name: 'service1',
            duration: 15,
            price: 100
        }, {
            id: 2,
            name: 'service2',
            duration: 16,
            price: 101
        },]
    }
};
const salonId = 1;
describe('SalonDetails shallow and passing store directly', () => {
    let store, container;
    const mockStore = configureMockStore();

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(<ConnectedSalonDetails store={store} match={{params: {}}}/>)
    });

    it('render SalonDetails component', () => {
        const wrapper = shallow(
            <SalonDetails match={{params: {}}}/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('render connected(SalonDetails) component', () => {
        expect(container.length).toEqual(1)
    });
});

describe('SalonDetails mount and wrapping in Provider', () => {
    let store, wrapper;
    const mockStore = configureMockStore();

    beforeEach(() => {
        store = mockStore(initialState);
        wrapper = mount(<Provider store={store}><ConnectedSalonDetails match={{params: {salonId}}}/></Provider>)
    });

    it('render connected(SalonDetails) component', () => {
        expect(wrapper.length).toEqual(1)
    });

    it('check Props matches with initialState', () => {
        store.subscribe(() => {
            expect(wrapper.find(ConnectedSalonDetails).prop('isLoadingSalon')).toEqual(initialState.salons.item.isLoading);
            expect(wrapper.find(ConnectedSalonDetails).prop('isLoadingServices')).toEqual(initialState.salonServices.isLoading);
            expect(wrapper.find(ConnectedSalonDetails).prop('salon')).toEqual(initialState.salons.item);
            expect(wrapper.find(ConnectedSalonDetails).prop('services')).toEqual(initialState.salonServices.services);

        });
    });

    it('check actions on dispatching', () => {
        expect(store.getActions()).toEqual([
            {type: Salons.SALON_REQUEST, payload: { salonId}},
            {type: SalonServices.SALON_SERVICES_REQUEST, payload: {salonId}},
        ]);
    });
});