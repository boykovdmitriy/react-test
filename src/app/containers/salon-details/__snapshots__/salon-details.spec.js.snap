// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalonDetails shallow and passing store directly render SalonDetails component 1`] = `<Spinner />`;
